File: packages/python/src/alumnium/server/accessibility/accessibility_tree_diff.py
from dataclasses import dataclass, field
from difflib import unified_diff


@dataclass
class AccessibilityTreeDiff:
    before_xml: str
    after_xml: str
    _diff: str = field(init=False, default="")

    def compute(self) -> str:
        if not self._diff:
            self._diff = self._format_as_git_diff()

        return self._diff

    def _format_as_git_diff(self) -> str:
        before_lines = self.before_xml.splitlines(keepends=True)
        after_lines = self.after_xml.splitlines(keepends=True)

        # Ensure last lines have newlines for consistent diff output
        if before_lines and not before_lines[-1].endswith("\n"):
            before_lines[-1] += "\n"
        if after_lines and not after_lines[-1].endswith("\n"):
            after_lines[-1] += "\n"

        diff = unified_diff(
            before_lines,
            after_lines,
            fromfile="before",
            tofile="after",
        )

        return "".join(diff).rstrip("\n")

File: packages/core/src/server/accessibility/accessibilityTreeDiff.ts
// @ts-expect-error -- TODO: Missing Python API
import { unifiedDiff } from "difflib";

export class AccessibilityTreeDiff {
  beforeXml: string;
  afterXml: string;
  #diff: string = "";

  constructor(beforeXml: string, afterXml: string) {
    this.beforeXml = beforeXml;
    this.afterXml = afterXml;
  }

  compute(): string {
    if (!this.#diff) {
      this.#diff = this.#formatAsGitDiff();
    }

    return this.#diff;
  }

  #formatAsGitDiff(): string {
    // @ts-expect-error -- TODO: Missing Python API
    const beforeLines = this.beforeXml.splitlines({ keepends: true });
    // @ts-expect-error -- TODO: Missing Python API
    const afterLines = this.afterXml.splitlines({ keepends: true });

    // Ensure last lines have newlines for consistent diff output
    const beforeLastIndex = beforeLines.length - 1;
    if (beforeLastIndex >= 0) {
      const beforeLastLine = beforeLines[beforeLastIndex];
      if (beforeLastLine !== undefined && !beforeLastLine.endsWith("\n")) {
        beforeLines[beforeLastIndex] = `${beforeLastLine}\n`;
      }
    }
    const afterLastIndex = afterLines.length - 1;
    if (afterLastIndex >= 0) {
      const afterLastLine = afterLines[afterLastIndex];
      if (afterLastLine !== undefined && !afterLastLine.endsWith("\n")) {
        afterLines[afterLastIndex] = `${afterLastLine}\n`;
      }
    }

    const diff = unifiedDiff(beforeLines, afterLines, {
      fromfile: "before",
      tofile: "after",
    });

    return diff.join("").replace(/\n$/, "");
  }
}
